### PR TITLE
Implements service factory

### PR DESCRIPTION
**Pull request by boyan-tonchev**
What?
Implements a service for dependency injection.

Why?
As developers, we needs uniform way to create services and inject them into another service.

How?
Implement a service which have to be used when an instance of another service is needed. It takes care of service creation and a service to have only one instance.

### CODE REVIEW

File: ontotext-yasgui-web-component/docs/developers-guide.md [hunk 1]
```diff
@@ -94,20 +94,18 @@ yasgui-tooltip tag have to wrap the element to which tooltip have to be appeared
 
 # Services
 
-Services are just plain javascript classes.
+Services are just plain javascript classes. 
 
-Services should be implemented as singletons. For example:
+There is a service factory which get care of service creation.
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/docs/developers-guide.md:99]: `which get care of` > `which is responsible for`

File: ontotext-yasgui-web-component/docs/developers-guide.md [hunk 2]
```diff
@@ -94,20 +94,18 @@ yasgui-tooltip tag have to wrap the element to which tooltip have to be appeared
 
 # Services
 
-Services are just plain javascript classes.
+Services are just plain javascript classes. 
 
-Services should be implemented as singletons. For example:
+There is a service factory which get care of service creation. 
+This factory have to be used everywhere a service is needed.
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/docs/developers-guide.md:100]: `This factory have to be` > `The factory must be`

File: ontotext-yasgui-web-component/src/services/service-factory.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import {EventService} from './event-service';
+
+/**
+ * The purpose of this @{see ServiceFactory} is to manage all other services used into the "ontotext-yasgui-web-component" components.
+ * It takes care for:
+ * <ul>
+ *   <li>building of service</li>
+ *   <li>every service to have only one instance (Singleton)</li>
+ *
+ * This factory have to be used everywhere a service is needed. Construction of an instance with "new" operator will brooke the Singleton of the service
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/service-factory.ts:11]: Same notes as in the developers-guide

File: ontotext-yasgui-web-component/src/services/service-factory.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      ServiceFactory.postConstruct(instance);
+      this._instances.set(type.name, instance);
+    }
+    return this._instances.get(type.name);
+  }
+
+  private static postConstruct<T>(instance: T) {
+    switch (instance.constructor.name) {
+      case "YasqeService":
+        (instance as YasqeService).init();
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/service-factory.ts:35]: The explicit YasqeService.init() invocation can now be removed and replaced with constructor initialization.

File: ontotext-yasgui-web-component/src/services/service-factory.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        break;
+      case "EventService":
+        (instance as EventService).hostElement = this._hostElement;
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/service-factory.ts:38]: This if this can be removed as well.

**Review by svilenvelikov — CHANGES_REQUESTED** [review]

File: ontotext-yasgui-web-component/src/services/yasgui/ontotext-yasgui-service.ts [hunk 6]
```diff
@@ -2,39 +2,40 @@ import {YasguiConfiguration} from '../../models/yasgui-configuration';
...
+    this.translationService = serviceFactory.get(TranslationService);
   }
 
   postConstruct(hostElement: HTMLElement, config: YasguiConfiguration): void {
 
-    OntotextYasguiServiceDefinition.initEditorTabs(hostElement, config);
-    OntotextYasguiServiceDefinition.initControlBar(hostElement, config);
-    OntotextYasguiServiceDefinition.initResultTabs(hostElement, config);
-    OntotextYasguiServiceDefinition.initButtonsStyling(hostElement, config);
-    OntotextYasguiServiceDefinition.updateTranslation(config);
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/yasgui/ontotext-yasgui-service.ts:20]: Why did you make all these instance methods? They are perfectly OK to be static.

Comment (boyan-tonchev) [reply ontotext-yasgui-web-component/src/services/yasgui/ontotext-yasgui-service.ts:20]: You are right. I will revert them.

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**